src/libsystemd/sd-device/device-util.c

```diff
@@ -113,3 +115,84 @@ bool device_match_parent(sd_device *device, Set *match_parent, Set *nomatch_pare
+                sd_device *device,
+                const char *field,
+                int (*func)(sd_device *dev, const char **s),
+                char ***strv) {
+
+        const char *s;
+        int r;
+
+        assert(device);
+        assert(field);
+        assert(func);
+        assert(strv);
+
+        r = func(device, &s);
+        if (r < 0 && r != -ENOENT)
+                log_debug_errno(r, "Failed to get device \"%s\" property, ignoring: %m", field);
+
+        if (r >= 0)
+                (void) strv_extendf(strv, "%s=%s", field, s);
+
+        return 0;
+}
+
+char** device_log_context_fields(sd_device *device) {
+        _cleanup_strv_free_ char **strv = NULL;
+        dev_t devnum;
+        int ifindex;
+        sd_device_action_t action;
+        uint64_t seqnum, diskseq;
+        int r;
+
+        assert(device);
+
+        (void) add_string_field(device, "SYSPATH", sd_device_get_syspath, &strv);
+        (void) add_string_field(device, "SUBSYSTEM", sd_device_get_subsystem, &strv);
+        (void) add_string_field(device, "DEVTYPE", sd_device_get_devtype, &strv);
+        (void) add_string_field(device, "DRIVER", sd_device_get_driver, &strv);
+        (void) add_string_field(device, "DEVPATH", sd_device_get_devpath, &strv);
+        (void) add_string_field(device, "DEVNAME", sd_device_get_devname, &strv);
+        (void) add_string_field(device, "SYSNAME", sd_device_get_sysname, &strv);
+        (void) add_string_field(device, "SYSNUM", sd_device_get_sysnum, &strv);
+
+        r = sd_device_get_devnum(device, &devnum);
+        if (r < 0 && r != -ENOENT)
+                log_debug_errno(r, "Failed to get device \"DEVNUM\" property, ignoring: %m");
+
+        if (r >= 0)
+                (void) strv_extendf(&strv, "DEVNUM="DEVNUM_FORMAT_STR, DEVNUM_FORMAT_VAL(devnum));
+
+        r = sd_device_get_ifindex(device, &ifindex);
+        if (r < 0 && r != -ENOENT)
+                log_debug_errno(r, "Failed to get device \"IFINDEX\" property, ignoring: %m");
+
+        if (r >= 0)
+                (void) strv_extendf(&strv, "IFINDEX=%i", ifindex);
+
+        r = sd_device_get_action(device, &action);
+        if (r < 0 && r != -ENOENT)
+                log_debug_errno(r, "Failed to get device \"ACTION\" property, ignoring: %m");
+
+        if (r >= 0)
+                (void) strv_extendf(&strv, "ACTION=%s", device_action_to_string(action));
+
+        r = sd_device_get_seqnum(device, &seqnum);
+        if (r < 0 && r != -ENOENT)
+                log_debug_errno(r, "Failed to get device \"SEQNUM\" property, ignoring: %m");
+
+        if (r >= 0)
+                (void) strv_extendf(&strv, "SEQNUM=%"PRIu64, seqnum);
+
+        r = sd_device_get_diskseq(device, &diskseq);
+        if (r < 0 && r != -ENOENT)
+                log_debug_errno(r, "Failed to get device \"DISKSEQ\" property, ignoring: %m");
+
+        if (r >= 0)
+                (void) strv_extendf(&strv, "DISKSEQ=%"PRIu64, diskseq);
+
+        return TAKE_PTR(strv);
```

Comment:
How about to cash the strv in sd-device object? Then, we can enable/disable log context frequently with almost no performance regression.

Reply:
See the other discussion, I don't think we'll be enabling/disabling the log context very frequently, so I'd like to leave this optimization until we find a use case where we're enabling/disabling the same log context very frequently.